rust-src/bulletproofs/src/set_membership_proof.rs

```diff
@@ -0,0 +1,734 @@
+        let mut rx_i = r_1[i];
+        rx_i.mul_assign(&x);
+        rx_i.add_assign(&r_0[i]);
+        rx.push(rx_i);
+    }
+    // Compute t(x)
+    // tx <- t_0 + t_1*x + t_2*x^2
+    let mut tx = t_0;
+    let mut tx_1 = t_1;
+    tx_1.mul_assign(&x);
+    tx.add_assign(&tx_1);
+    let mut tx_2 = t_2;
+    tx_2.mul_assign(&x_sq);
+    tx.add_assign(&tx_2);
+    // Compute the blinding t_x_tilde
+    // t_x_tilde <- z^2*v_rand + t_1_tilde*x + t_2_tilde*x^2
+    let mut tx_tilde = z_sq;
+    tx_tilde.mul_assign(v_rand);
+    let mut tx_s1 = t_1_tilde;
+    tx_s1.mul_assign(&x);
+    tx_tilde.add_assign(&tx_s1);
+    let mut tx_s2 = t_2_tilde;
+    tx_s2.mul_assign(&x_sq);
+    tx_tilde.add_assign(&tx_s2);
+    // Compute blinding e_tilde
+    // e_tilde <- a_tilde + s_tilde * x
+    let mut e_tilde = s_tilde;
+    e_tilde.mul_assign(&x);
+    e_tilde.add_assign(&a_tilde);
+    // append tx, tx_tilde, e_tilde to transcript
+    transcript.append_message(b"tx", &tx);
+    transcript.append_message(b"tx_tilde", &tx_tilde);
+    transcript.append_message(b"e_tilde", &e_tilde);
+
+    // Part 5: Inner product proof for tx = <lx,rx>
+    // get challenge w from transcript
+    let w: C::Scalar = transcript.challenge_scalar::<C, _>(b"w");
+    // get generator q
+    let Q = B.mul_by_scalar(&w);
+    // compute scalars c such that c*H = H', that is H_prime_scalars = (1, y^-1,..,
+    // y^-(n-1))
+    let y_inv = match y.inverse() {
+        Some(inv) => inv,
+        None => return Err(ProverError::DivisionError),
+    };
+    let H_prime_scalars = z_vec(y_inv, 0, n);
+    // compute inner product proof
+    let proof =
+        prove_inner_product_with_scalars(transcript, &G, &H, &H_prime_scalars, &Q, &lx, &rx);
+
+    // return set membership proof
+    if let Some(ip_proof) = proof {
+        Ok(SetMembershipProof {
+            A,
+            S,
+            T_1,
+            T_2,
+            tx,
+            tx_tilde,
+            e_tilde,
+            ip_proof,
+        })
+    } else {
+        Err(ProverError::InnerProductProofFailure)
+    }
+}
+
+/// Error messages detailing why proof verification failed
+#[derive(Debug, PartialEq, Eq)]
+pub enum VerificationError {
+    /// The set size must be representable by an unsigned 64-bit integer
+    SetTooLarge,
+    /// The set must have a size of a power of two
+    SetSizeNotPowerOfTwo,
+    /// The length of `gens` was less than `|the_set|`
+    NotEnoughGenerators,
+    /// The consistency check for `t_0` failed
+    InconsistentT0,
+    /// Choice of randomness led to verification failure
+    DivisionError,
+    /// Inner product proof verification failed
+    IPVerificationError,
+}
+
+/// This function verifies a set membership proof, i.e. a proof of knowledge
+/// of value v that is in a set S and that is consistent
+/// with a commitment V to v. The arguments are
+/// - `transcript` - the random oracle for Fiat Shamir
+/// - `the_set` - the set as a vector
```

Comment:
Please make it explicit here that this only works for sets of size power of two. In bold :) 

Alternatively just pad in this function (and in the prover). Nobody outside of this crate should use this verifier and prover directly.

Or add another function that pads + verifies and pads + proves.

Reply:
Added padding directly into prove and verify.